cirq-core/cirq/protocols/unitary_protocol_test.py

```diff
@@ -201,6 +235,23 @@ def test_decompose_and_get_unitary():
     np.testing.assert_allclose(_strat_unitary_from_decompose(DummyComposite()), np.eye(1))
     np.testing.assert_allclose(_strat_unitary_from_decompose(OtherComposite()), m2)
 
+    np.testing.assert_allclose(
+        _strat_unitary_from_decompose(GateThatAllocatesAQubit()),
+        GateThatAllocatesAQubit._target_unitary,
+    )
```

Comment:
I would suggest creating a new parameterized test for `test_gate_allocates_new_qubits` and then let all the parameterized gates have a common interface to get the `gate._target_unitary()`. I'd also suggest making `GateThatAllocatesAQubit` a parameterized test where we apply `Z ** self.theta` instead of `Z` to the ancilla. 

I imagine we'd like to cover many different "classes" of gates that can allocate new qubits (eg: clean ancilla, dirty ancilla, recursive allocation on <= 4 qubits, >= 4 qubits etc.)